mesonbuild/coredata.py

```diff
@@ -594,6 +594,12 @@ def process_new_compilers(self, lang: str, comp, cross_comp, env):
             # Unlike compiler and linker flags, preprocessor flags are not in
             # compiler_options because they are not visible to user.
             preproc_flags = shlex.split(preproc_flags)
+            k = lang + '_args'
+            if lang in ('c', 'cpp', 'objc', 'objcpp') and k in env.properties[for_machine]:
+                # `c_args` in the cross file are used, like CPPFLAGS but *not*
+                # CFLAGS, for tests. this is weird, but how it was already
+                # implemented. Hopefully a new version of #3916 fixes it.
+                preproc_flags = stringlistify(env.properties[for_machine][k])
```

Comment:
@nirbheek this does do the status quo. for `[properties]` in the cross file, and `[properties]` in the machine file would it be added. I suspect you rather have an independent `c_preproc_args` like `c_link_args` to keep the same separation as `CPPFLAGS` and `<LANG>FLAGS`.